File: kits/php/src/Registry.php
<?php

declare(strict_types=1);

namespace IronMcp;

/**
 * Self-discovery of ironmcp servers. File-backed, with a cross-process O_EXCL lock around every
 * read-modify-write (closes the lost-update TOCTOU, invariant #9), pid-liveness pruning on read
 * (lazy GC, invariant #10), and an XDG path (not ~/.loqu8). The `namespace` field on each entry
 * keeps it estate-wide rather than Loqu8-only.
 *
 * On-disk format is byte-compatible with the Dart `IronMcpRegistry`
 * (kits/dart/lib/src/registry.dart): a single JSON object at
 * $XDG_RUNTIME_DIR|$XDG_STATE_HOME|~/.local/state + /ironmcp/registry.json, keyed by entry id,
 * each value a {@see RegistryEntry} shape. A Dart server and a PHP server share ONE file.
 *
 * PHP has O_EXCL via fopen(..., 'x'): the open fails (returns false) if the file already exists, an
 * atomic create-or-fail — the same primitive Dart's File.create(exclusive: true) uses.
 */
final class Registry
{
    private readonly string $dir;

    /** @var callable(int):bool */
    private $isPidAlive;

    /**
     * @param string|null            $dir           registry directory; defaults to the XDG path
     * @param (callable(int):bool)|null $isPidAlive  liveness probe; defaults to a POSIX/proc check
     * @param float                  $lockTimeout   seconds to wait for the lock before best-effort proceed
     * @param float                  $staleLockAfter seconds after which a stale lock is stolen
     */
    public function __construct(
        ?string $dir = null,
        ?callable $isPidAlive = null,
        private readonly float $lockTimeout = 3.0,
        private readonly float $staleLockAfter = 30.0,
    ) {
        $this->dir = $dir ?? self::defaultDir();
        $this->isPidAlive = $isPidAlive ?? self::pidAliveDefault(...);
    }

    private function file(): string
    {
        return $this->dir . '/registry.json';
    }

    private function lockFile(): string
    {
        return $this->dir . '/registry.json.lock';
    }

    public static function defaultDir(): string
    {
        $env = getenv();
        $base = $env['XDG_RUNTIME_DIR']
            ?? $env['XDG_STATE_HOME']
            ?? (($env['HOME'] ?? '.') . '/.local/state');

        return $base . '/ironmcp';
    }

    /** Register (or replace) an entry. Locked read-modify-write. */
    public function register(RegistryEntry $entry): void
    {
        $this->withLock(function () use ($entry): void {
            $map = $this->read();
            // toArray() already casts capabilities to an object; casting the entry itself to an
            // object keeps the whole value object-shaped so a rewrite never flips {} to [].
            $map->{$entry->id} = (object) $entry->toArray();
            $this->write($map);
        });
    }

    /** Remove an entry by id. Locked read-modify-write. */
    public function unregister(string $id): void
    {
        $this->withLock(function () use ($id): void {
            $map = $this->read();
            unset($map->{$id});
            $this->write($map);
        });
    }

    /**
     * Live servers, pruning any whose pid is dead (and rewriting the file if it pruned). A
     * hard-killed process is cleaned up lazily on the next reader's scan, since its own unregister
     * never ran (invariant #10).
     *
     * @return list<RegistryEntry>
     */
    public function discover(): array
    {
        $live = [];
        $this->withLock(function () use (&$live): void {
            $map = $this->read();
            $pruned = false;
            foreach (get_object_vars($map) as $key => $raw) {
                if (!is_object($raw)) {
                    unset($map->{$key});
                    $pruned = true;
                    continue;
                }
                $entry = RegistryEntry::fromArray((array) $raw);
                if (($this->isPidAlive)($entry->pid)) {
                    $live[] = $entry;
                } else {
                    unset($map->{$key});
                    $pruned = true;
                }
            }
            if ($pruned) {
                $this->write($map);
            }
        });

        return $live;
    }

    /**
     * Read the registry as an object tree. Decoding to objects (not associative arrays) preserves
     * the JSON distinction between {} and [] across a read-modify-write — critical for cross-kit
     * compatibility, since a Dart consumer reads `capabilities` as a Map and a stray [] would throw.
     */
    private function read(): \stdClass
    {
        $file = $this->file();
        if (!is_file($file)) {
            return new \stdClass();
        }
        $txt = @file_get_contents($file);
        if ($txt === false || trim($txt) === '') {
            return new \stdClass();
        }
        try {
            $decoded = json_decode($txt, false, 512, \JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return new \stdClass(); // corrupt/unreadable: start fresh rather than crash
        }

        return $decoded instanceof \stdClass ? $decoded : new \stdClass();
    }

    /**
     * Atomic write: encode to a unique tmp file on the same filesystem, then rename over the target.
     * An empty top-level object encodes as {} (matching Dart), never [].
     */
    private function write(\stdClass $map): void
    {
        $this->ensureDir();
        $json = json_encode($map, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_UNESCAPED_UNICODE);
        if ($json === false) {
            $json = '{}';
        }
        $tmp = \sprintf('%s.tmp.%d.%s', $this->file(), getmypid() ?: 0, bin2hex(random_bytes(6)));
        file_put_contents($tmp, $json, \LOCK_EX);
        if (!@rename($tmp, $this->file())) {
            @unlink($tmp);
        }
    }

    /** @param callable():void $body */
    private function withLock(callable $body): void
    {
        $this->ensureDir();
        $lock = $this->lockFile();
        $deadline = microtime(true) + $this->lockTimeout;
        $acquired = false;
        while (true) {
            // fopen(..., 'x') is O_EXCL|O_CREAT: fails (false) if the lock already exists.
            $fh = @fopen($lock, 'x');
            if ($fh !== false) {
                fclose($fh);
                $acquired = true;
                break;
            }
            // A crashed holder can leave a stale lock — steal it past staleLockAfter.
            $mtime = @filemtime($lock);
            if ($mtime !== false && (microtime(true) - $mtime) > $this->staleLockAfter) {
                @unlink($lock);
                continue;
            }
            if (microtime(true) >= $deadline) {
                break; // proceed best-effort
            }
            usleep(5000); // 5ms
        }
        try {
            $body();
        } finally {
            if ($acquired) {
                @unlink($lock);
            }
        }
    }

    private function ensureDir(): void
    {
        if (!is_dir($this->dir)) {
            @mkdir($this->dir, 0o700, true);
        }
    }

    /** Default liveness probe: POSIX kill(0) when available, else /proc, else fail open. */
    private static function pidAliveDefault(int $pid): bool
    {
        if (\function_exists('posix_kill')) {
            if (posix_kill($pid, 0)) {
                return true;
            }
            // EPERM (1) means the process exists but is owned by another user — still alive.
            return \defined('PHP_INT_MAX') && posix_get_last_error() === 1;
        }
        if (is_dir('/proc')) {
            return is_dir("/proc/{$pid}");
        }

        return true; // fail open — never prune a live entry we cannot verify
    }
}

File: kits/python/src/ironmcp/registry.py
"""Self-discovery of ironmcp servers over a shared, language-neutral registry file.

A Dart iCE, a Python ``*light`` server, and a Node scarlight all read and write the SAME
discovery fabric. The on-disk format is byte-for-byte compatible with
``kits/dart/lib/src/registry.dart`` (READ it): a JSON object keyed by entry-id, each value
an entry of snake_case fields, an ``O_EXCL`` lockfile around every read-modify-write, and
pid-liveness pruning on read.

Entry deliberately carries NO hand-kept tool list — a consumer enumerates a server's tools
via ``tools/list`` on its port (loqu8 invariant #3: the list that drifted from 6 to 66).
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Callable, Optional

__all__ = ["IronMcpEntry", "IronMcpRegistry"]


def _now_iso() -> str:
    """ISO-8601 in UTC, matching Dart's ``DateTime.now().toUtc().toIso8601String()``."""
    return datetime.now(timezone.utc).isoformat()


@dataclass
class IronMcpEntry:
    """A live ironmcp server's registration. Language-neutral JSON (snake_case)."""

    id: str
    namespace: str
    pid: int
    host: Optional[str] = None
    port: Optional[int] = None
    transport: Optional[str] = None
    version: Optional[str] = None
    code_sha: Optional[str] = None
    capabilities: dict = field(default_factory=dict)
    started_at: str = field(default_factory=_now_iso)

    def to_json(self) -> dict:
        # Key order + optional-field omission mirror the Dart toJson exactly.
        out: dict = {"id": self.id, "namespace": self.namespace, "pid": self.pid}
        if self.host is not None:
            out["host"] = self.host
        if self.port is not None:
            out["port"] = self.port
        if self.transport is not None:
            out["transport"] = self.transport
        if self.version is not None:
            out["version"] = self.version
        if self.code_sha is not None:
            out["code_sha"] = self.code_sha
        out["capabilities"] = self.capabilities
        out["started_at"] = self.started_at
        return out

    @staticmethod
    def from_json(j: dict) -> "IronMcpEntry":
        return IronMcpEntry(
            id=j["id"],
            namespace=j["namespace"],
            pid=int(j["pid"]),
            host=j.get("host"),
            port=j.get("port"),
            transport=j.get("transport"),
            version=j.get("version"),
            code_sha=j.get("code_sha"),
            capabilities=dict(j.get("capabilities") or {}),
            started_at=j.get("started_at") or _now_iso(),
        )


def _default_dir() -> str:
    """XDG_RUNTIME_DIR else XDG_STATE_HOME else ~/.local/state, then ``/ironmcp``.

    Identical to the Dart ``_defaultDir`` so both kits resolve the same file.
    """
    env = os.environ
    base = (
        env.get("XDG_RUNTIME_DIR")
        or env.get("XDG_STATE_HOME")
        or os.path.join(env.get("HOME", "."), ".local", "state")
    )
    return os.path.join(base, "ironmcp")


def _pid_alive_default(pid: int) -> bool:
    """True iff ``pid`` is a live process. Fails OPEN — never prune what we can't verify."""
    if pid <= 0:
        return False
    try:
        os.kill(pid, 0)
        return True
    except ProcessLookupError:
        return False
    except PermissionError:
        return True  # exists but owned by another user
    except OSError:
        return True  # fail open


class IronMcpRegistry:
    """File-backed registry with a cross-process O_EXCL lock, pid pruning, and an XDG path."""

    def __init__(
        self,
        *,
        dir: Optional[str] = None,
        is_pid_alive: Optional[Callable[[int], bool]] = None,
        lock_timeout: float = 3.0,
        stale_lock_after: float = 30.0,
    ) -> None:
        self._dir = dir if dir is not None else _default_dir()
        self._is_pid_alive = is_pid_alive or _pid_alive_default
        self.lock_timeout = lock_timeout
        self.stale_lock_after = stale_lock_after

    @property
    def _file(self) -> str:
        return os.path.join(self._dir, "registry.json")

    @property
    def _lock_file(self) -> str:
        return os.path.join(self._dir, "registry.json.lock")

    def register(self, entry: IronMcpEntry) -> None:
        def body() -> None:
            data = self._read()
            data[entry.id] = entry.to_json()
            self._write(data)

        self._with_lock(body)

    def unregister(self, id: str) -> None:
        def body() -> None:
            data = self._read()
            data.pop(id, None)
            self._write(data)

        self._with_lock(body)

    def discover(self) -> list[IronMcpEntry]:
        """Live servers, pruning any whose pid is dead (and rewriting the file if it
        pruned). A hard-killed process is cleaned up lazily on the next reader's scan,
        since its own ``unregister`` never ran (invariant #10)."""
        live: list[IronMcpEntry] = []

        def body() -> None:
            data = self._read()
            pruned = False
            for key in list(data.keys()):
                entry = IronMcpEntry.from_json(dict(data[key]))
                if self._is_pid_alive(entry.pid):
                    live.append(entry)
                else:
                    del data[key]
                    pruned = True
            if pruned:
                self._write(data)

        self._with_lock(body)
        return live

    # --- internals -------------------------------------------------------------

    def _read(self) -> dict:
        try:
            if not os.path.exists(self._file):
                return {}
            with open(self._file, encoding="utf-8") as f:
                txt = f.read()
            if not txt.strip():
                return {}
            return dict(json.loads(txt))
        except Exception:
            return {}  # corrupt/unreadable: start fresh rather than crash

    def _write(self, data: dict) -> None:
        os.makedirs(self._dir, exist_ok=True)
        tmp = f"{self._file}.tmp.{os.getpid()}.{time.time_ns()}"
        with open(tmp, "w", encoding="utf-8") as f:
            json.dump(data, f, indent=2)
        os.replace(tmp, self._file)  # atomic on the same filesystem

    def _with_lock(self, body: Callable[[], None]) -> None:
        os.makedirs(self._dir, exist_ok=True)
        deadline = time.monotonic() + self.lock_timeout
        acquired = False
        while True:
            try:
                # O_EXCL: atomic create-or-fail, the cross-process mutex (matches Dart's
                # File.create(exclusive: true)).
                fd = os.open(self._lock_file, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
                os.close(fd)
                acquired = True
                break
            except FileExistsError:
                # A crashed holder can leave a stale lock — steal it past stale_lock_after.
                try:
                    age = time.time() - os.path.getmtime(self._lock_file)
                    if age > self.stale_lock_after:
                        os.unlink(self._lock_file)
                        continue
                except OSError:
                    pass
                if time.monotonic() > deadline:
                    break  # proceed best-effort
                time.sleep(0.005)
        try:
            body()
        finally:
            if acquired:
                try:
                    os.unlink(self._lock_file)
                except OSError:
                    pass
